src/eval.py

```diff
@@ -0,0 +1,164 @@
+from typing import Dict, List, Optional, Sequence, Union, cast
+
+import numpy as np
+import pandas as pd
+import torch
+from einops import repeat
+from sklearn.base import BaseEstimator, clone
+from sklearn.ensemble import RandomForestClassifier
+from sklearn.linear_model import LogisticRegression
+from sklearn.metrics import f1_score, precision_score, recall_score
+from torch.utils.data import DataLoader
+from tqdm import tqdm
+
+from .dataset import WorldCerealLabelledDataset
+from .presto import Presto, PrestoFineTuningModel
+from .utils import DEFAULT_SEED, device
+
+
+class WorldCerealEval:
+    name = "WorldCerealCropland"
+    threshold = 0.5
+
+    def __init__(self, train_data: pd.DataFrame, val_data: pd.DataFrame, seed: int = DEFAULT_SEED):
+        self.seed = seed
+        self.train_ds = train_data
+        self.val_ds = val_data
+
+    @staticmethod
+    def _mask_to_batch_tensor(
+        mask: Optional[np.ndarray], batch_size: int
+    ) -> Optional[torch.Tensor]:
+        # TODO: This function should be replaced by a real mask,
+        # returned by the dataloader
+        if mask is not None:
+            return repeat(torch.from_numpy(mask).to(device), "t c -> b t c", b=batch_size).float()
+        return None
+
+    @torch.no_grad()
+    def finetune_sklearn_model(
+        self,
+        dl: DataLoader,
+        pretrained_model,
+        mask: Optional[np.ndarray] = None,
+        models: List[str] = ["Regression", "Random Forest"],
+    ) -> Union[Sequence[BaseEstimator], Dict]:
+        for model_mode in models:
+            assert model_mode in ["Regression", "Random Forest"]
+        pretrained_model.eval()
+
+        encoding_list, target_list = [], []
+        for x, y, dw, latlons, month in tqdm(dl, desc="Computing embeddings"):
+            x, dw, latlons, y, month = [t.to(device) for t in (x, dw, latlons, y, month)]
+            batch_mask = self._mask_to_batch_tensor(mask, x.shape[0])
+            target_list.append(y.cpu().numpy())
+            with torch.no_grad():
+                encodings = (
+                    pretrained_model.encoder(
+                        x, dynamic_world=dw.long(), mask=batch_mask, latlons=latlons, month=month
+                    )
+                    .cpu()
+                    .numpy()
+                )
+                encoding_list.append(encodings)
+        encodings_np = np.concatenate(encoding_list)
+        targets = np.concatenate(target_list)
+        if len(targets.shape) == 2 and targets.shape[1] == 1:
+            targets = targets.ravel()
+
+        fit_models = []
+        model_dict = {
+            "Regression": LogisticRegression(
```

Comment:
for my understanding, what's the purpose of the logistic regression task?

Reply:
This is just legacy code - its implemented for free in the [main presto codebase](https://github.com/nasaharvest/presto/blob/8d8563ab680160edfded1dbf153c0f8b0b866c93/presto/eval/eval.py#L160), and I thought we might want to run this as a quick approximation of finetuning Presto. This is why I left it in rather than deleting it.